### PR TITLE
restrict ability to use phone for 2fa

### PR DESCRIPTION
**Pull request by mjriley**
## Summary
Context on the JIRA ticket: https://dimagi-dev.atlassian.net/browse/SAAS-12515. Submitting early to get approval for the method.

## Safety Assurance

- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

Several test suites added:
- settings/tests/test_forms
- users/tests/test_users
- settings/tests/test_views

### QA Plan

The testing is comprehensive and I've run this back and forth through local testing. QA feels redundant.

### Safety story

I've done local testing to verify these changes work with existing 2fa users, those without it, those with phone backup tokens, etc.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/settings/forms.py [hunk 1]
```diff
@@ -140,9 +139,9 @@ def clean_token(self):
 
 class HQTwoFactorMethodForm(MethodForm):
 
-    def __init__(self, **kwargs):
-        super(HQTwoFactorMethodForm, self).__init__(**kwargs)
-        if not settings.ALLOW_PHONE_AS_DEFAULT_TWO_FACTOR_DEVICE:
+    def __init__(self, *, allow_phone, **kwargs):
```
Comment (dannyroberts) [inline corehq/apps/settings/forms.py:142]: I might change this variable to be called `allow_phone_2fa` or `allow_phone_as_two_factor_device` so there's less doubt for someone reading the code.

File: corehq/apps/settings/views.py [hunk 2]
```diff
@@ -372,6 +394,13 @@ def dispatch(self, request, *args, **kwargs):
         # this is only here to add the login_required decorator
         return super(TwoFactorSetupView, self).dispatch(request, *args, **kwargs)
 
+    def get_form_kwargs(self, step=None):
+        kwargs = super().get_form_kwargs(step)
+        if step == 'method':
+            kwargs.setdefault('allow_phone', _user_can_use_phone(self.request.couch_user))
```
Comment (dannyroberts) [inline corehq/apps/settings/views.py:400]: same comment about `allow_phone` name

File: corehq/apps/settings/views.py [hunk 3]
```diff
@@ -66,6 +67,20 @@
...
 
+def _user_can_use_phone(user):
+    if not settings.ALLOW_PHONE_AS_DEFAULT_TWO_FACTOR_DEVICE:
+        return False
+
+    domains = (Domain.get_by_name(domain) for domain in user.domains)
+
+    # The reason we iterate through domains, rather than fetch them all at once (there is a view to do so)
+    # is due to concerns about scale. Most users belong to one or a few domains, so iteration isn't expensive.
+    # For users that DO belong to many domains, I'm working off the assumption that most of them are for
```
Comment (orangejenny) [inline corehq/apps/settings/views.py:78]: I agree this is a reasonable assumption.

**Review by millerdev — APPROVED** [review]

**Comment by dannyroberts** [comment]
Looks like there's a (fairly minor) conflict to resolve

**Pull request merged**